### PR TITLE
Issue #63: Distinguish between new and changed lines

### PR DESCRIPTION
**Pull request by HuGanghui**
Issue #63: Distinguish between new and changed lines

this PR should wait PR #61 and PR https://github.com/java-diff-utils/java-diff-utils/pull/90（has been merged)

### CODE REVIEW

**Comment by HuGanghui** [comment]
set add option false : https://huganghui.github.io/patch-filter-63/add-false-DiffReport/
set add option true: https://huganghui.github.io/patch-filter-63/add-true-DiffReport/

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java [hunk 1]
```diff
@@ -157,14 +171,114 @@ private String getFileName(List<String> singlePatch) {
...
-            lineRange.add(targetChunk.getPosition());
-            lineRange.add(targetChunk.getPosition() + targetChunk.getLines().size());
-            lineRangeList.add(lineRange);
+            final List<Integer> removePositionList = getRemovePositionList(patch, i);
+            final List<Integer> addPositionList = getAddPositionList(patch, i);
+            if (add) {
+                final List<Integer> newPositionList = new ArrayList<>();
+                int addIndex = 0;
+                int removeIndex = 0;
+                while (addIndex < addPositionList.size()) {
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:195]: It is hard to understand what is going on in this loop. It looks like the code is duplicated a lot. Please split the logic into separate meaningful methods.

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java [hunk 2]
```diff
@@ -60,6 +60,11 @@
      */
     private boolean optional;
 
+    /**
+     * Control if only consider added lines in file.
+     */
+    private boolean add;
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:66]: Please refer to the initial doc and call it `strategy`.

File: pom.xml [hunk 3]
```diff
@@ -26,7 +26,7 @@
         <dependency>
             <groupId>io.github.java-diff-utils</groupId>
             <artifactId>java-diff-utils</artifactId>
-            <version>4.5</version>
+            <version>4.8-SNAPSHOT</version>
```
Comment (timurt) [inline pom.xml:29]: Is it temporal `version`? Could you provide snapshot url or it is local build?

Comment (timurt) [reply pom.xml:29]: How can I execute this branch on local?

Comment (HuGanghui) [reply pom.xml:29]: @timurt you can git clone https://github.com/java-diff-utils/java-diff-utils, and using mater branch, and run `mvn clean install ` to get `4.8-SNAPSHOT`

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 1 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 2 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java [hunk 4]
```diff
@@ -60,6 +61,11 @@
      */
     private boolean optional;
 
+    /**
+     * Control if only consider added lines in file.
+     */
+    private boolean strategy;
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:67]: Why `boolean`? Please read the doc and check values this field is supposed to have

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java [hunk 5]
```diff
@@ -114,6 +129,7 @@ private void loadPatchFile(String patchFileName) throws CheckstyleException {
                         new SuppressionPatchFilterElement(fileName, lineRangeList);
                 filters.addFilter(element);
             }
+            System.out.println();
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:132]: Remove

File: src/main/resources/patchConfig.xml [hunk 6]
```diff
@@ -18,7 +18,8 @@
     </module>
 
     <module name="com.puppycrawl.tools.checkstyle.filters.SuppressionPatchFilter">
-        <property name="file" value="${checkstyle.patchfilter.patch}"/>
+        <property name="file" value="/Users/hgh/Documents/GitHub/patch-filters/DiffReport/patch.txt"/>
```
Comment (rdiachenko) [inline src/main/resources/patchConfig.xml:21]: Why did you remove a placeholder?

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java [hunk 7]
```diff
@@ -159,16 +175,190 @@ private String getFileName(List<String> singlePatch) {
 
     private List<List<Integer>> getLineRange(Patch<String> patch) {
         final List<List<Integer>> lineRangeList = new ArrayList<>();
+        int totalAddNum = 0;
+        int totalRemoveNum = 0;
         for (int i = 0; i < patch.getDeltas().size(); i++) {
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:180]: This for loop should not take more than 15 lines. Please, explain your logic in words. Give as much details as possible. Do this in the issue comment and put here a reference. Let's move discussion into the issue.

File: src/main/resources/patchConfig.xml (hunk 6 above)
Comment (HuGanghui) [reply src/main/resources/patchConfig.xml:21]: done, sorry, it is an oversight.

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 5 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:132]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 4 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 7 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:179]: ok, link is here: https://github.com/checkstyle/patch-filters/issues/63#issuecomment-655266516

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 4 above)
Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java]: Please clarify where did you find `strategy = "all"` and `"added".equals(strategy)` strategies in the doc?

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 7 above)
Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:179]: The logic looks very complex, not maintainable and tricky to test, even checkstyle complains and you ignore those complains which is not good:
```
    <suppress checks="CyclomaticComplexity" files="SuppressionPatchFilter\.java"/>
    <suppress checks="ExecutableStatementCount" files="SuppressionPatchFilter\.java"/>
    <suppress checks="JavaNCSS" files="SuppressionPatchFilter\.java"/>
```
I'm checking your comment in the issue, thanks for details you provided but it might take some time. In the meantime, could you please think:
1. How can you simplify the logic?
2. Are there other ways to update https://github.com/java-diff-utils/java-diff-utils/pull/90/files which would simplify the aforementioned logic?

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 4 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java]: >Based on the aforementioned cases the following strategies should be considered during implementation:
> line - consider only those violations which were occurred in new code lines only
>context - consider a wider context when new code introduces violations outside of new/changed lines, but let's >presume it is on the same branch in AST but on upper level.
>column - distinguish between existing and new violations on the changed code lines

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 7 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:179]: @rdiachenko I simplify the logic using inner class and make a new PR, please review https://github.com/checkstyle/patch-filters/pull/70

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 4 above)
Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java]: Please show me in this sentence from the doc where did you find `changed lines both belong to line strategy`:
> line - consider only those violations which were occurred in new code lines only

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 7 above)
Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:179]: It is not a simplification. You just hid it within the inner class.

Where are the answers to my two questions above?

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 4 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java]: > * context - consider a wider context when new code introduces violations outside of new/changed lines, 
according to this sentence, context is about violations outside of new/changed lines. So I think changed lines belongs to line strategy, and 

> * line - consider only those violations which were occurred in new code lines only

this line strategy definition is not proper.

I understand this example belongs to line strategy:

image

how about this one, I also think them belong to line strategy, because we violations are introduced inside of new/changed lines, not satisfy context's condition——`context - consider a wider context when new code introduces violations outside of new/changed lines`:

image

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java (hunk 7 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilter.java:179]: @rdiachenko I've thought about it, but I haven't come up with a better way to update https://github.com/java-diff-utils/java-diff-utils/pull/90/files.

**Comment by rdiachenko** [comment]
We are going to move to JGit as java-diff-utils has a lot of restrictions. This PR is closed in favor of https://github.com/checkstyle/patch-filters/issues/76
